crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -121,11 +121,13 @@ where
     }
 
     /// Schedule a state change that replaces the active state with the given state.
-    /// This will fail if there is a scheduled operation, pending transition, or if the given
-    /// `state` matches the current state
+    /// This will fail if there is a scheduled operation or a pending transition.
+    ///
+    /// If `state` is the same as the current one, this does nothing.
+    /// Use [`restart`](#method.restart) to trigger `on_exit` and `on_enter` for the current state.
     pub fn set(&mut self, state: T) -> Result<(), StateError> {
         if self.current_state == state {
-            return Err(StateError::AlreadyInState);
+            return Ok(());
```

Comment:
This will now return `Ok` even when a transition to another state is pending.
Placing this after the other check would fix that :)

Reply:
Nice catch, thanks